x-pack/test/observability_ai_assistant_api_integration/tests/chat/chat.spec.ts

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import expect from '@kbn/expect';
+import { MessageRole, type Message } from '@kbn/observability-ai-assistant-plugin/public';
+import getPort from 'get-port';
+import http from 'http';
+import { PassThrough } from 'stream';
+import { FtrProviderContext } from '../../common/ftr_provider_context';
+
+export default function ApiTest({ getService }: FtrProviderContext) {
+  const supertest = getService('supertest');
+
+  const CHAT_API_URL = `/internal/observability_ai_assistant/chat`;
+
+  const messages: Message[] = [
+    {
+      '@timestamp': new Date().toISOString(),
+      message: {
+        role: MessageRole.System,
+        content: 'You are a helpful assistant',
+      },
+    },
+    {
+      '@timestamp': new Date().toISOString(),
+      message: {
+        role: MessageRole.User,
+        content: 'Good morning!',
+      },
+    },
+  ];
+
+  describe('Chat', () => {
+    let requestHandler = (
+      request: http.IncomingMessage,
+      response: http.ServerResponse<http.IncomingMessage> & { req: http.IncomingMessage }
+    ) => {};
+
+    let connectorId: string;
+
+    before(async () => {
+      const port = await getPort({ port: getPort.makeRange(9000, 9100) });
+
+      http
+        .createServer((request, response) => {
+          requestHandler(request, response);
+        })
+        .listen(port);
+
+      const response = await supertest
+        .post('/api/actions/connector')
+        .set('kbn-xsrf', 'foo')
+        .send({
+          name: 'OpenAI',
+          connector_type_id: '.gen-ai',
+          config: {
+            apiProvider: 'OpenAI',
+            apiUrl: `http://localhost:${port}`,
+          },
+          secrets: {
+            apiKey: 'my-api-key',
+          },
+        })
+        .expect(200);
+
+      connectorId = response.body.id;
+    });
+
+    it("returns a 4xx if the connector doesn't exist", async () => {
+      await supertest
+        .post(CHAT_API_URL)
+        .set('kbn-xsrf', 'foo')
+        .send({
+          messages,
+          connectorId: 'does not exist',
+        })
+        .expect(404);
+    });
+
+    it('returns a streaming response from the server', async () => {
+      const NUM_RESPONSES = 5;
+
+      requestHandler = (request, response) => {
+        response.writeHead(200, {
+          'Content-Type': 'text/event-stream',
+          'Cache-Control': 'no-cache',
+          Connection: 'keep-alive',
+        });
+
+        new Array(NUM_RESPONSES).fill(undefined).forEach(() => {
+          response.write(`data: {}`);
+        });
+
+        response.write('data: [DONE]');
+        response.end();
+      };
+
+      await Promise.race([
```

Comment:
What benefit is a Promise.race adding here ?

Reply:
It makes sure the test doesn't hang if there's a bug. The timeout will reject the promise after 5s and fail the test. It's easy to have stuff indefinitely hanging when working with streams.

Reply:
Makes sense, given we make one API call (SuperTest) which makes another API call which we are mocking, it's good to have a timeout 👍🏼